### ISSUE TITLE
POST /api/v1/user/에서 IntegrityError가 발생 

### ISSUE THREAD

**Issue by 201432420** [wafflestudio/rookies#227]
#196 Issue, #201 Issue와 동일하게, POST /api/v1/user/를 시도할 때 IntegrityError가 뜨는 문제를 겪고 있습니다. 

저는 UserSerializer에서 role을 다음과 같이 정의하고, 

role = serializers.ChoiceField(['participant', 'instructor'], read_only=True)

user/serializers.py의 create()에서 role을 확인하고 어떤 값이냐에 따라 해당하는 ParticipantProfile, InstructorProfile을 생성하려고 했습니다. 코드는 다음과 같습니다. 

<img width="912" alt="Screen Shot 2020-10-03 at 5 02 29 PM" src="https://user-images.githubusercontent.com/52521749/94986575-e204cd00-059a-11eb-915e-342f5671edb0.png">

DB에서 auth_user, authtoken_token에 object가 생성이 된 것을 확인했지만, user_participantprofile에는 생성이 되지 않았습니다. #196 이슈에서 제시한대로 auth_user와 authtoken table을 truncate를 해보았지만 문제는 해결되지 않았습니다. 

<img width="584" alt="Screen Shot 2020-10-03 at 4 55 26 PM" src="https://user-images.githubusercontent.com/52521749/94986580-e8934480-059a-11eb-8162-9e245638aa70.png">

print()를 써서 확인해본 결과, ParticipantProfile.objects.create(user=user) 이후에는 실행이 안 되는 것으로 보이는데, 어떤 부분이 문제인지 잘 모르겠습니다. objects.create()는 object를 생성하고 저장하는 메소드인 것으로 알고 있습니다. 과제를 두 차례 정도 처음부터 갈아엎었으나 문제를 해결하지 못했습니다 ㅠㅠ 어떻게 하면 이 문제를 해결할 수 있을까요? 

+ 추가 질문을 드리자면 위 코드에서 hasattr을 통해 university 값 등이 존재하는지 확인하는 것보다 더 나은 방법이 있는지 궁금합니다. 다른 방법으로 context=self.context를 통해 데이터를 넘기는 것에 대해 들어보았지만 이해가 잘 가지 않았습니다. 제가 찾은 [documentation](https://www.django-rest-framework.org/api-guide/serializers/#including-extra-context)에서 간략한 설명이 나와있었지만 serializers.py의 create() 경우에서처럼 nested representation을 하려는 경우에는 어떻게 적용될 수 있는지 잘 모르겠습니다.

**Comment by gyusang** [comment]
IntegrityError는 데이터베이스와 관련있는 에러입니다.
ParticipantProfile model을 어떻게 정의하셨나요?

한편, validated_data는 serializer의 field만을 포함합니다. 따라서 university를 확인하기 위해서는 serializer에 university라는 추가적인 field를 정의해야 합니다.

**Comment by 201432420** [comment]
다음과 같이 정의했습니다.

<img width="713" alt="Screen Shot 2020-10-03 at 6 22 16 PM" src="https://user-images.githubusercontent.com/52521749/94988064-7d9b3b00-05a5-11eb-9ae5-b74f05b38290.png">

그리고 validated_data에 대해서 설명해주셔서 감사합니다. 다만 추가적인 질문이 있는데, 

<img width="621" alt="Screen Shot 2020-10-03 at 6 39 09 PM" src="https://user-images.githubusercontent.com/52521749/94988387-c3f19980-05a7-11eb-9630-738f670921b1.png">

UserSerializer에서 위와 같이 "university" 등의 field들이 정의되어 있는데도 불구하고 validated_data를 프린트하면 username, email, password에 대한 정보만 나오는 이유는 무엇인가요?print(validated_data.pop('university'))를 시도하면 KeyError: 'university'라는 오류가 뜨네요 ㅠㅠ

**Comment by gyusang** [comment]
ParticipantProfile은 잘 정의된 것 같은데, IntegrityError의 원인은 잘 모르겠습니다..

Debug Toolbar를 이용해서 어떤 SQL query가 보내지는지 확인해보는 것은 어떨까요?

university field가 read_only=True로 설정되어 있습니다. read_only인 경우에는 serialize할 때만 필드가 사용되고, update나 create를 위해 deserialize할 때는 사용되지 않습니다.

이 경우에는 write_only=True로 설정하는 게 적절해 보입니다.

**Comment by 201432420** [comment]
네, Debug Toolbar를 이용해 SQL query를 확인해보겠습니다. 조언 주셔서 감사합니다. university의 경우도 지적해주신 사항대로 고쳤습니다.
